src/Microsoft.AspNetCore.Mvc.Testing/WebApplicationFactory.cs

```diff
@@ -0,0 +1,330 @@
+    {
+        private TestServer _server;
+        private IWebHostBuilder _builder;
+        private IList<HttpClient> _clients = new List<HttpClient>();
+
+        /// <summary>
+        /// <para>
+        /// Creates an instance of <see cref="WebApplicationFactory{TEntryPoint}"/>. This factory can be used to
+        /// create a <see cref="TestServer"/> instance using the MVC application defined by <typeparamref name="TEntryPoint"/>
+        /// and one or more <see cref="HttpClient"/> instances used to send <see cref="HttpRequestMessage"/> to the <see cref="TestServer"/>.
+        /// The <see cref="WebApplicationFactory{TEntryPoint}"/> will find the entry point class of <typeparamref name="TEntryPoint"/>
+        /// assembly and initialize the application by calling <c>IWebHostBuilder CreateWebHostBuilder(string [] args)</c>
+        /// on <typeparamref name="TEntryPoint"/>.
+        /// </para>
+        /// <para>
+        /// This constructor will infer the application content root path by searching for a
+        /// <see cref="WebApplicationFactoryContentRootAttribute"/> on the assembly containing the functional tests with
+        /// a key equal to the <typeparamref name="TEntryPoint"/> assembly <see cref="Assembly.FullName"/>.
+        /// In case an attribute with the right key can't be found, <see cref="WebApplicationFactory{TEntryPoint}"/>
+        /// will fall back to searching for a solution file (*.sln) and then appending <typeparamref name="TEntryPoint"/> asembly name
+        /// to the solution directory. The application root directory will be used to discover views and content files.
+        /// </para>
+        /// <para>
+        /// The application assemblies will be loaded from the dependency context of the assembly containing
+        /// <typeparamref name = "TEntryPoint" />.This means that project dependencies of the assembly containing
+        /// <typeparamref name = "TEntryPoint" /> will be loaded as application assemblies.
+        /// </para>
+        /// </summary>
+        public WebApplicationFactory()
+        {
+        }
+
+        /// <summary>
+        /// Gets the <see cref="TestServer"/> created by this <see cref="WebApplicationFactory{TEntryPoint}"/>.
+        /// </summary>
+        public TestServer Server => _server;
+
+        /// <summary>
+        /// Gets the <see cref="WebApplicationFactoryClientOptions"/> used by <see cref="CreateClient()"/>.
+        /// </summary>
+        public WebApplicationFactoryClientOptions ClientOptions { get; } = new WebApplicationFactoryClientOptions();
+
+        /// <summary>
+        /// The <see cref="IWebHostBuilder"/> used to create the <see cref="TestServer"/>.
+        /// </summary>
+        public IWebHostBuilder WebHostBuilder
+        {
+            get
+            {
+                if (_server != null)
+                {
+                    return null;
```

Comment:
This is still weird... I would hope no one would use the `?.` in their code for this because that would mean they're ignoring a serious bug in their code. I feel this is an anti-pattern in the making.

An alternative that I think avoids the problems is Separation of Concerns. This type is used for construction. When you want the server, you call `.CreateServer()`, and that returns a server to you, from which you can get clients, make requests, etc.